test/libyul/YulName.cpp

```diff
@@ -0,0 +1,86 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <boost/test/unit_test.hpp>
+
+#include <libyul/Dialect.h>
+#include <libyul/YulName.h>
+#include <libyul/backends/evm/EVMDialect.h>
+
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace solidity::yul::test
+{
+
+BOOST_AUTO_TEST_SUITE(YulName)
+
+BOOST_AUTO_TEST_CASE(repository_generate_labels_for_derived_types)
+{
+	Dialect dialect{};
+	YulNameRepository nameRepository;
+	auto const test1 = nameRepository.defineName("test1");
+	BOOST_CHECK(!nameRepository.isDerivedName(test1));
+	auto const test1_1 = nameRepository.deriveName(test1);
+	BOOST_CHECK(nameRepository.isDerivedName(test1_1));
+	auto const test1_2 = nameRepository.deriveName(test1);
+	BOOST_CHECK(nameRepository.isDerivedName(test1_2));
+	auto const test1_2_1 = nameRepository.deriveName(test1_2);
+	BOOST_CHECK(nameRepository.isDerivedName(test1_2_1));
+	auto const test2 = nameRepository.defineName("test2_1");
+	auto const test2_1 = nameRepository.deriveName(test2);
+	auto const test3 = nameRepository.defineName("test3");
+	auto const test3_1 = nameRepository.deriveName(test3);
+	BOOST_CHECK(test1 != test1_1);
+	BOOST_CHECK(test1 < test1_1);
+	nameRepository.generateLabels({test1, test1_1, test1_2, test1_2_1, test2_1, test3, test3_1}, {"test1"});
+
+	// marking test1 as invalid means that all labels get bumped up by one
+	BOOST_CHECK(nameRepository.labelOf(test1) == "test1_1");
```

Comment:
Oh man it's been a while. IIRC this should just test that labels that are marked as off-limits are also treated as such. For example when you have externally defined symbols from the point of view of inline assembly. But I agree, ideally this would already fail elsewhere :)

Reply:
Maybe you could `assert` that the handles passed to `generateLabels` are `derived` names?
Is there anyone else could add their opinion on this?
Apart from this, I think this should be OK to merge.

Reply:
I don't think this should be asserted against, it might very well be that we have labels that never change (although unlikely with disambiguator running beforehand) and then you'd have the original id in ast. Which should go into generate labels and be ignored essentially, unless it's one that is deemed invalid :)